File: client-android/app/src/androidTest/java/stasis/client_android/mocks/MockAnalyticsClient.kt
package stasis.client_android.mocks

import stasis.client_android.lib.telemetry.analytics.AnalyticsClient
import stasis.client_android.lib.telemetry.analytics.AnalyticsEntry
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Companion.map
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

class MockAnalyticsClient(private val result: Try<Unit>) : AnalyticsClient {

    private val sentRef: AtomicInteger = AtomicInteger(0)

    private val lastEntryRef: AtomicReference<AnalyticsEntry?> = AtomicReference(null)

    override suspend fun sendAnalyticsEntry(entry: AnalyticsEntry): Try<Unit> =
        result.map {
            sentRef.incrementAndGet()
            lastEntryRef.set(entry)
        }

    val sent: Int
        get() = sentRef.get()

    val lastEntry: AnalyticsEntry?
        get() = lastEntryRef.get()

    companion object {
        operator fun invoke(): MockAnalyticsClient =
            MockAnalyticsClient(result = Try.Success(Unit))
    }
}

File: layers/src/test/scala/stasis/layers/telemetry/analytics/MockAnalyticsClient.scala
package stasis.layers.telemetry.analytics

import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

import scala.concurrent.Future

import org.apache.pekko.Done

class MockAnalyticsClient(result: Future[Done]) extends AnalyticsClient {
  import scala.concurrent.ExecutionContext.Implicits.global

  private val sentRef: AtomicInteger = new AtomicInteger(0)

  private val lastEntryRef: AtomicReference[Option[AnalyticsEntry]] = new AtomicReference(None)

  override def sendAnalyticsEntry(entry: AnalyticsEntry): Future[Done] =
    result
      .map { _ =>
        val _ = sentRef.incrementAndGet()
        lastEntryRef.set(Some(entry))
        Done
      }

  def sent: Int = sentRef.get()

  def lastEntry: Option[AnalyticsEntry] = lastEntryRef.get()
}

object MockAnalyticsClient {
  def apply(): MockAnalyticsClient =
    new MockAnalyticsClient(result = Future.successful(Done))

  def MockAnalyticsClient(result: Future[Done]): MockAnalyticsClient =
    new MockAnalyticsClient(result = result)
}
